driver-sync/src/main/com/mongodb/client/internal/Crypt.java

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.internal;
+
+import com.mongodb.MongoInternalException;
+import com.mongodb.MongoSocketReadException;
+import com.mongodb.ServerAddress;
+import com.mongodb.client.model.vault.DataKeyOptions;
+import com.mongodb.client.model.vault.EncryptOptions;
+import com.mongodb.crypt.capi.MongoCrypt;
+import com.mongodb.crypt.capi.MongoCryptContext;
+import com.mongodb.crypt.capi.MongoDataKeyOptions;
+import com.mongodb.crypt.capi.MongoExplicitEncryptOptions;
+import com.mongodb.crypt.capi.MongoKeyDecryptor;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonBinary;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.bson.RawBsonDocument;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static com.mongodb.crypt.capi.MongoCryptContext.State;
+
+class Crypt implements Closeable {
```

Comment:
This is the state machine, which makes use of MongoCrypt (the libmongocrypt wrapper) and the various utility classes to do the work.